[Yaml] deprecate parsing mappings without keys

xabbuh opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        |

xabbuh committed:
[Yaml] deprecate parsing mappings without keys

xabbuh commented on src/Symfony/Component/Yaml/Tests/InlineTest.php:
```diff
@@ -420,4 +420,22 @@ public function testVeryLongQuotedStrings()
+    public function testTheEmptyStringIsAValidMappingKey()
+    {
+        $this->assertSame(array('' => 'foo'), Inline::parse('{ "": foo }'));
+    }
```
The two tests above were already passing without this patch. I just added them to prevent future regressions.

xabbuh replied:
Not sure if we shouldn't just add these two tests in `2.7` now.

stof replied:
👍 for adding them in an older version, to prevent regressions on them when doing bugfixes in older branches

xabbuh replied:
see #21650

stof commented:
what is the existing behavior for a mapping without key ?

stof commented:
if the parser is not broken currently for such case, and so accepts the input (even though it is not compliant), we should rather deprecate it instead, to be consistent with our previous changes where we give users time to update their YAML files

xabbuh commented:
The parser behaviour currently is inconsistent since some changes we made in 3.2. However, you are right that we should deprecate it first on `master` (#21647 will make sure that the wrong behaviour is restored in the `3.2` branch).

Status: Needs work

xabbuh commented:
I updated here to deprecate the possibility to omit mapping keys and changed the target branch to `master`. Please not that this contains the changes from #21647 and needs to be merged afterwards.

Status: Needs Review

stof commented:
the other PR is now merged (including in master)

xabbuh commented:
rebased

stof commented:
👍

fabpot commented on UPGRADE-3.3.md:
```diff
@@ -119,6 +119,8 @@ Workflow
 Yaml
 ----
 
+ * Omitting the key of a mapping is deprecated and will lead to a `ParseException` in Symfony 4.0.
```
`and throws a`?

xabbuh replied:
updated

fabpot commented:
Thank you @xabbuh.
